### DISCUSSION TITLE
git clone get error:The remote end hung up unexpectedly

### DISCUSSION THREAD

**Discussion by fredgan** [community/community#39210] [Repositories]
### Select Topic Area

Product Feedback

### Body

Hi,
When I try to `git clone https://github.com/ClickHouse/ClickHouse`, I get the errors:

```
# git clone https://github.com/ClickHouse/ClickHouse
Cloning into 'ClickHouse'...
error: RPC failed; HTTP 403 curl 22 The requested URL returned error: 403 Forbidden
fatal: The remote end hung up unexpectedly
```

While other repos won't get this error.

Can anybody help me? Why is `ClickHouse` so special? How can I clone this repo?

**Comment by DSDmark** [comment]
Hi @fredgan,

I'm not sure, but you can try these steps:-

### Try to change your protocol, then try.

You can use ssh or curl, which you want. Here is with **ssh**.

- ```git remote add origin git@github.com:ClickHouse/ClickHouse.git```

If you do that then not sure, but it requires **ssh key**. And also In some case it's not adding your origin, then try that first.

- ```git remote rm origin```

> The good point of SSH comparing HTTPS is that you can use one ssh key for several repositories and git hostings. I use one key for github, bitbucket and gitlab in despite of different usernames for every hosting. It saves a lot of time for fetch & push operations

> SSH I also see a performance increase while cloning.

---

If you not get your answer with above solutions then you try this.

- Try to change your Wi-Fi network from broadband to your mobile hotspot. 

---

- You can also use ```git config ssh.postBuffer 524288000``` if posting over ssh instead of http.

For some time, I guess you need to add **global** tag in this method.

- ``` git config --global http.postBuffer 100000000```.

---

Not sure with that but I think nothing is better for some things so, you could try this.
- ```git gc --aggressive```

---

Again, If you not get your answer with above solutions, then you try this.

- delete .git/ from your project
clone the remote repo to some new location, like your desktop:

- ```git clone https://github.com/ClickHouse/ClickHouse.git```
move ```.git/``` from the new location into the old location
re-commit and push your changes

---

### Why this happening (I just copy past that section. There is no perfect answer why this happening).

> This is not an answer per se but more of an observation that may help others. I found that this error pops up occasionally on high-latency networks (I have to use a Satellite dish for internet access for example). The speed of the network is fine, but the latency can be high. Note: The problem only exists in certain scenarios, but # editing this answer  wait....I have not determined what the pattern is.

> Temporary mitigation:
I switched networks—I moved to a slower, but lower latency cell network (my phone used as a hotspot)—and the problem disappeared. Note that I can only do this itermittently because my cell connectivity is also intermittent. Plus, the bandwidth usage adds costs. I'm also lucky that I have this option available to me. Not everyone does.

> I'm sure there is some configuration setting somewhere that makes git—or ssh or curl or whatever times out first—more tolerant of such networks, but I don't know what it is.

**Reply by fredgan** [reply]
Thanks for your suggestion. But they don't work.
I am asking my IT workmate to research this problem now. If any progress, I will report it here.

**Reply by ghost** [reply]
It worked for me thanks.

**Reply by JavierAbrego** [reply]
same here it fixed my issue, thanks @DSDmark !

**Reply by DevNinja254** [reply]
The problem is the buffer: just increase its space
for http => git config http.postBuffer 524288000;
     ssh  => git config ssh.postBuffer 524288000;`

**Reply by AditSoni** [reply]
@DevNinja254  Dude thanks, this was the problem.

**Comment by airtower-luna** [comment]
The repository URL is missing `.git` at the end, it should be:
```
git clone https://github.com/ClickHouse/ClickHouse.git
```
It might be that there's some other issue, too, but fixing that removes at least one problem.

**Reply by fredgan** [reply]
Thanks for your suggestion. 
The suffix `.git` is unnecessary. Both are OK.

**Comment by fredgan** [comment]
Result: IT said that it's limited by the network to upload package bigger than 1M.

ClickHouse has thousands of branches and tags. I don't know what does `git clone` do in the detail. Maybe it would upload all the branch names.

**Comment by airtower-luna** [comment]
A clone will *download* all the repository data from the repository you're cloning. But for a large repository that may involve a lot of back and forth of discovering what's available and actually retrieving it, and those requests add up. Sounds like IT has a questionable view of network management. :confused: 

You can try to reduce the amount of data `git clone` retrieves by telling it not to fetch *everything*, e.g. with the [`--single-branch` option](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---no-single-branch) or by doing a shallow clone. Obviously I can't say if that'll be enough.

The proper fix would be to talk to whoever is in charge of that policy and get them to change it.

**Comment by mitchiemt11** [comment]
It might help someone:
In my case, I was getting the same issue when trying to push to my repo and the following commands saved my day:

```
git config --global http.postBuffer 1048576000
git config --global http.lowSpeedLimit 0
git config --global http.lowSpeedTime 999999
```

**Reply by Khydn** [reply]
OMG that work for me !

**Reply by notaf0e** [reply]
Wow thanks
